pkg/threescale/backend/backend.go

```diff
@@ -419,22 +419,34 @@ func (b *Backend) handleFlushAuthorization(apps []*handledApp) []*handledApp {
 // In both of these cases, a successful call instead of an error will result in a change in the mentioned state
 func (b *Backend) handleFlushCacheUpdate(apps []*handledApp) {
 	for _, app := range apps {
+
+		cacheKey := app.snapshot.getCacheKey()
+		getCachedApp := func() (*Application, bool) {
+			return b.cache.Get(cacheKey)
+		}
+
 		// handle case (1)
 		if app.reportingErr && app.authErr {
-			// nothing to do here only continue on
 			// todo add logging
-			continue
-		}
-		cacheKey := app.snapshot.getCacheKey()
-		//now that we have handled the dual error case, we know we need to
-		// read from the cache and modify the entry so we take a write lock
-		cachedApp, ok := b.cache.Get(cacheKey)
-		if !ok {
+			// we failed to report this so place the old report back on the queue
+			b.queue.append(&app.snapshot)
+
+			// we need to check if this report contains elapsed time period so we can handle it correctly
+			cachedApp, ok := getCachedApp()
+			if !ok {
+				continue
+			}
+
+			cachedApp.handleElapsedTimePeriod(app.snapshot.timestamp, time.Now())
 			continue
 		}
 
 		// handle case (4)
 		if !app.reportingErr && app.authErr {
+			cachedApp, ok := getCachedApp()
+			if !ok {
+				continue
+			}
```

Comment:
this is now executed in both sides of the `if`, so can extract

Reply:
ok, was trying to call this JIT before taking the lock, not at the beginning of each iteration?

Reply:
since you take the lock after getting the cached app in both cases it does not matter if it is right before or not, since any window big or small before locking should not be important